Add bookings

NurievLinar opened the pull request.

NurievLinar committed:
Item. создал поля

NurievLinar committed:
ТЗ 13. Каркас приложения ShareIt.

NurievLinar committed:
ItemRequest.java : удалил неиспользуемые импорты.

NurievLinar committed:
ТЗ 13. Исправления приложения ShareIt.
DuplicateEmailException - сделал наследование от RuntimeException;
ItemNotFoundException - сделали наследование от RuntimeException;
UserNotFoundException - сделали наследование от RuntimeException;
В соответствующих методах удалил указывание throws;
ItemRepositoryImpl - исправил item1 на currentItem;
ItemRepositoryImpl - удалил .equals(true);

NurievLinar committed:
Удалил неиспользуемые импорты, мешающие прохождению тестов на GitHub.

NurievLinar committed:
Исправления по Checkstyle, для прохождения тестов на GitHub.

NurievLinar committed:
ТЗ 14. Добавление бронирования.

NurievLinar committed:
ItemServiceImpl. Исправление недочётов.

sergeymamonov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,64 @@
+
+    private final BookingService bookingService;
+
+    public BookingController(BookingService bookingService) {
```
Не обязательно писать конструктор руками, если можно воспользоваться аннотацией Lombok @AllArgsConstructor? Аналогично и для других классов, где можно воспользоваться этой аннотацией

NurievLinar replied:
Исправлено

sergeymamonov commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+            case REJECTED:
+                bookings = bookingRepository.findAllByBooker_IdAndStatus(userId, Status.REJECTED);
+                break;
+        }
```
Лучше всегда делать ветку default на случай, если мы получим статус, который нам неизвестен, чтобы можно было кинуть ошибку или что-то еще сделать в такой ситуации.

Ниже аналогично.

NurievLinar replied:
Добавил ветку default.
Проверка статуса реализована методом validateState().

sergeymamonov commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,168 @@
+            itemDtoList.add(itemDto);
+        }
+        for (ItemDto itemDto : itemDtoList) {
+            List<Comment> commentList = commentRepository.findAllByItem_Id(itemDto.getId());
```
Не очень хорошо делать запросы к БД в цикле. Это большая нагрузка на БД. Затратно по времени из-за передачи по сети.

NurievLinar replied:
Исправлено.

sergeymamonov commented on src/main/java/ru/practicum/shareit/item/ItemServiceImpl.java:
```diff
@@ -0,0 +1,168 @@
+                    .collect(Collectors.toList());
+            itemDto.setComments(commentDtos);
+            Sort sortDesc = Sort.by(Sort.Direction.DESC, "end");
+            Optional<Booking> lastBooking = bookingRepository.findTop1BookingByItem_IdAndEndIsBeforeAndStatusIs(
```
Здесь и в строке 120 аналогично, запросы к БД в цикле.

NurievLinar replied:
Исправлено.

sergeymamonov commented on src/main/java/ru/practicum/shareit/user/exception/ValidationException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class ValidationException extends RuntimeException {
```
Это исключение где-то используется?

NurievLinar replied:
Исправлено

sergeymamonov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,53 @@
+CREATE TABLE IF NOT EXISTS users
+(
+    id    BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+    name  VARCHAR NOT NULL,
```
В скриптах sql размер varchar должен быть ограничен.

NurievLinar replied:
Исправлено.

NurievLinar committed:
ТЗ 14. Исправление замечаний после ревью.
BookingController. Конструктор удалён. Добавлена аннотация @AllArgsConstructor.
BookingMapper. Добавлены мапперы для lastBooking, nextBooking.
ItemServiceImpl. Исправлено обращение к БД в цикле.
ItemMapper. Добавлен маппер.
Удалены неиспользуемые классы.

NurievLinar committed:
ТЗ 14. Исправление замечаний GitHab
